Language support

scambra opened the pull request:
I have added support to look for movies in different language than english. It's on languages branch, I didn't want to push to master until it's completed and reviewed.

It adds another page on settings, languages, to add a language code (language-COUNTRY, ISO 639-1 and ISO 3166-1), as used by TheMovieDB. Each language code has language names, which are names for the language to require them in title, to consider movie is in requested language.

Movies can be edited to change language, then it queries themoviedb to look for titles in that language (it also changes plot with translated plot in language if available), and allows user to pick one title (sometimes multiple titles are provided in alternative titles and translations). English title is saved in new column, english_title.

When searching for torrents, 2 queries are sent, with translated title and english title. Language's names are removed from ignore words list, and one name is required to be in torrent's name to compare torrent's title with english_title.

If user request movie's original language, then english title is not saved, so movie is looked only with original title, and language's names are not required to be in the title, although they are removed from ignore words list.

@barbequesauce @ellnic thoughts?

barbequesauce commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 20
           

Complexity increasing per file
==============================
- static/js/settings/languages.js  4
- core/library.py  3
         

Clones added
============
- core/ajax.py  4
- core/library.py  1
- static/js/library/status.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754&bid=20290178)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+    display: inline-block;
+    text-align: center;
+}
+#languages tr .names{
```
Codacy found an issue: [Expected single space before "{" (block-opening-brace-space-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce replied:
Codacy found an issue: [Expected empty line before rule (rule-empty-line-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
```
Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on core/ajax.py:
```diff
@@ -861,6 +881,37 @@ def update_metadata(self, imdbid, tmdbid=None):
 
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def movie_metadata(self, imdbid, tmdbid=None, language=None):
```
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+    var blanks = false;
+
+// Categories
+    var required_fields = ["code", "names"];
```
Codacy found an issue: ['required_fields' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on core/app.py:
```diff
@@ -80,6 +80,7 @@ def defaults(self):
     search_template = Template(filename='templates/settings/search.html', module_directory=core.MAKO_CACHE)
     quality_template = Template(filename='templates/settings/quality.html', module_directory=core.MAKO_CACHE)
     categories_template = Template(filename='templates/settings/categories.html', module_directory=core.MAKO_CACHE)
+    languages_template = Template(filename='templates/settings/languages.html', module_directory=core.MAKO_CACHE)
```
Codacy found an issue: [Mako templates allow HTML/JS rendering by default and are inherently open to XSS attacks. Ensure variables in all templates are properly sanitized via the 'n', 'h' or 'x' flags (depending on context). For example, to HTML escape the variable 'data' do ${ data |h }.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
+    display: inline-block;
+    text-align: center;
```
Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+/*global each, category_template, $categories_form */
```
Codacy found an issue: ['category_template' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+                     });
+}
+
+function _get_settings(){
```
Codacy found an issue: ['_get_settings' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+    $categories_form = $("form#languages tbody");
+});
+
+function add_language(event){
```
Codacy found an issue: ['add_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
+    display: inline-block;
```
Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+            return;
+        }
+
+        langs[code] = names;
```
Codacy found an issue: [Generic Object Injection Sink](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+
+function delete_language(event, button){
+    event.preventDefault();
+    var $tr = $(button).closest('tr');
```
Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
```
Codacy found an issue: [Expected single space before "{" (block-opening-brace-space-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/library/status.js:
```diff
@@ -602,6 +613,47 @@ function update_metadata(event, elem, imdbid, tmdbid){
     });
 }
 
+function change_language(event, elem, imdbid, tmdbid){
```
Codacy found an issue: ['change_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+/*global each, category_template, $categories_form */
+window.addEventListener("DOMContentLoaded", function(){
+    language_template = document.querySelector("template#language_template").innerHTML;
```
Codacy found an issue: ['language_template' is not defined.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+
+function add_language(event){
+    event.preventDefault();
+    var $new_language = $(language_template);
```
Codacy found an issue: ['language_template' is not defined.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/js/settings/languages.js:
```diff
@@ -0,0 +1,52 @@
+    $new_language.find("[data-toggle=\"tooltip\"]").tooltip();
+}
+
+function delete_language(event, button){
```
Codacy found an issue: ['delete_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

barbequesauce commented on static/css/settings/languages.css:
```diff
@@ -0,0 +1,10 @@
+    text-align: center;
+}
+#languages tr .names{
+    width: 69%;
```
Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

ellnic commented:
Nice job, Sergio :-) This is a much needed feature. If it’s working as expected, I don’t see any reason why it can’t be merged to master straight away.

scambra commented:
I had to change arguments for search method, it needs more movie's info such as english title and language, so I changed to get whole movie dict, and changed every method calling this. That's why I didn't push to master directly, it could break something, although I didn't see any issue in last few days. Maybe someone else could test it, no need to use languages, but ensure it still works as expected for english.

Here is one screenshot of UI changes in movie window
![image](https://user-images.githubusercontent.com/20515/94131008-b6c50400-fe5d-11ea-8c89-759c585b656d.png)

I haven't added language to add movie window, I can add it later

scambra commented:
@barbequesauce @ellnic do you want to test this branch? or is it ok to merge into master?

barbequesauce commented:
Yeah sorry I've been running it since Saturday - no real issues - I'm good with it, great work thank you!

ellnic commented:
Sorry for the delay in replying. I'm missing a lot of notifications at the moment (massively busy at work right now). October is always pretty busy so you might have to nag me if it's urgent.
